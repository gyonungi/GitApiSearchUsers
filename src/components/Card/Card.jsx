import { useSelector } from "react-redux";
import s from "./Card.module.css"
import { Link } from "react-router-dom";
import { useState } from "react";


const Card = ({date,displayusers}) => { 
  const [isShown, setIsShown] = useState(false);
  const handleClick = () => {
    setIsShown(current => !current);
  }
  const handlehref = () =>{
    window.location.href = `${date?.html_url}`
  }
  return (
    <>
      <ul  id={s.card}>
        <li className={s.personal}>
        <div className={s.avatar}>
         <img  src={date?.avatar_url} alt={date?.login} />
      </div>
      <li className={s.nameHolder}>{date?.login}
      </li>
        </li>
        <li className={s.info}>
     <li data-testid="list-item"> Repositories:{date?.repos_url?.length} </li>
         <button className={s.othBut} onClick={handleClick}>Подробности</button>
      </li>
      {isShown && (
      <li className={s.followers}>
        <Link onClick={handlehref} >Профиль на GitHub</Link> 
        <span> Followers: {date?.followers_url?.length} </span> 
        <span> Following: {date?.following_url?.length} </span> 
        <span> Stars: {date?.starred_url?.length} </span> 
         </li>
          )}
         </ul>
    </>
  );
};

export default Card;
