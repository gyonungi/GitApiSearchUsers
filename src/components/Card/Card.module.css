#card {
    font-family: "Segoe UI";
    margin: 12vh auto;
    border-radius: 5px;
    overflow: hidden;
    width: 350px;
    padding: 15px 25px;
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 3px 1px -2px rgba(0, 0, 0, 0.12), 0 1px 5px 0 rgba(0, 0, 0, 0.2);
}

.personal {
    display: flex;
    flex-direction: column;
    align-items: center;
}
.nameHolder  {
    font-size: 30px;
    margin: 0;
    padding-right: 20px;
}
.nameHolder h1 {
    font-size: 30px;
    margin: 0;
    padding-right: 20px;
}



.info {
    display: flex;
    padding: 20px 25px;
    font-weight: lighter;
    color: #3b3a3a;
    font-size: 20px;
    flex-wrap: wrap;
    align-content: center;
    justify-content: space-between;
}
.info h1 {
    font-size: 30px;
    margin: 0;
    padding-right: 20px;
}
.info h1 {
    font-size: 30px;
    margin: 0;
    padding-right: 20px;
}

.info li {
    padding-left: 55px;
    padding-bottom: 10px;
}

.avatar img {
    border-radius: 40px;
    width: 100px;
    height: 100px;
    border: 5px solid #fff;
    box-shadow: 0 4px 5px 0 rgba(0, 0, 0, 0.14), 0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.3);
    margin-right: 25px;
}

.skills {
    margin-bottom: 5px;
}

.skills span {
    display: inline-block;
    height: 32px;
    font-size: 13px;
    font-weight: 500;
    color: rgba(0, 0, 0, 0.6);
    line-height: 32px;
    padding: 0 12px;
    border-radius: 16px;
    background-color: #e4e4e4;
    margin-bottom: 5px;
    margin-right: 5px;
    
}

.followers {
    background-color: #f8f9f9;
    margin: 0px -25px;
    padding: 13px;
    text-align: center;
    color: #6e6e6e;
    border-top: 1px solid #eaeaea;
    display: grid;
    justify-content: center;
}

.followers  {
    font-weight: lighter;
    margin: 0;
    font-size: 15px;
}
li {
    list-style-type: none;
    /* Убираем маркеры */
  }
.animate-change {
    animation-name: slide-in;
    animation-duration: 0.3s;
}
.othBut {
display: block;
background-color: #009ee4;
padding: 2px 5px 2px 5px;
text-decoration: none;
text-align: center;
margin: 10px -25px 10px 70px;
line-height: 40px;
border: 1px solid #009ee4;
color: #fff;
font-weight: bold;
cursor: pointer;
}
.othBut:hover {
    background-color: #009ee4;
 }

