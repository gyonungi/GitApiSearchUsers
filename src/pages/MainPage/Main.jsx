import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link, NavLink } from "react-router-dom";
import s from "./Main.module.css";
import Card from "../../components/Card/Card";
import ReactPaginate from "react-paginate";
const Main = () => {
  const [inputValue, setInputValue] = useState("");
  const [isSorted, setisSorted] = useState(false);
  const [users, setUsers] = useState([]);

  const handleClick = () => {
    setisSorted((current) => !current);
  };
  /*   const dispatch = useDispatch(); */
  const getUsers = async (value) => {
    const response = await fetch(
      `https://api.github.com/search/users?q=${value}`,
      {headers: {'Content-Type': 'application/json'}, method: "GET"}
  );
    const users = await response.json();
    setUsers(users);
  };
  useEffect(() => {
    getUsers();
  }, []);
  const objectToArray = (data) => {
    return Object.keys(data).map((key) => {
      return {
        id: key,
        ...data,
      };
    });
  };
  /*   function searchGit(e) {
    e.preventDefault();
    dispatch(getUsers(inputValue));
  } */
  //INPUT
  const handleInputChange = (event) => {
    let inputValue = event.target.value;
    setInputValue(inputValue);
    getUsers(inputValue);
  };
  let gitUser = objectToArray(users);
  console.log(gitUser?.[2]?.items);

  //SORT
  const [sortuser, setSortUser] = useState(null);
  function numAscending() {
    const SortUser = gitUser?.[2]?.items;
    SortUser?.sort((a, b) => a?.repos_url?.length - b?.repos_url?.length);
  return  setSortUser(gitUser);
  }
  const [sortusers, setSortUsers] = useState(null);
  function numDescending() {
    gitUser?.[2]?.items?.sort(
      (a, b) => b?.repos_url?.length - a?.repos_url?.length
    );
    return  setSortUsers(gitUser);
  }

  //PAGINATINO
  const [page, setPage] = useState(0);
  const usersPerPage = 5;
  const numberOfusersVistited = page * usersPerPage;
  const displayusers = gitUser?.[2]?.items
    ?.slice(numberOfusersVistited, numberOfusersVistited + usersPerPage)  
    .map((item) => {
      return <Card key={item.id} date={item} />;
    });
  const totalPages = Math.ceil(gitUser?.[2]?.items?.length / usersPerPage);
  const changePage = ({ selected }) => {
    setPage(selected);
  };
  return (
    <>
      <h2 className={s.mainH2}>GitHub Search üîç</h2>
      <div className={s.mainSearch}>
        <form
          /* onSubmit={(e) => searchGit(e)} */
          className={s.searchForm}
          action="#"
        >
          <input
          data-testid="value-elem"
            className={s.searchText}
            placeholder="–ü–æ–∏—Å–∫ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è"
            name="search"
            type="text"
            value={inputValue}
            onChange={handleInputChange}
          />
        </form>
      </div>
      <div className={s.DivSort}> 
       <p>–°–æ—Ä—Ç–∏—Ä–æ–≤–∫–∞:</p> 
       <div>
      <button data-testid="Asc-btn" className={s.Asc} onClick={numAscending}>–í–Ω–∏–∑</button>
      <button data-testid="Desc-btn" className={s.Desc}onClick={numDescending}>–í–≤–µ—Ä—Ö</button>
      </div>
      </div>
      <div className={s.DivCards}>
        {users ? (displayusers): (<p>–ù–µ –Ω–∞–π–¥–µ–Ω –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å</p>) }
        <ReactPaginate
          previousLabel={"Previous"}
          nextLabel={"Next"}
          pageCount={totalPages}
          onPageChange={changePage}
          containerClassName={s.navigationButtons}
          previousLinkClassName={s.previousButton}
          nextLinkClassName={s.nextButton}
          disabledClassName={s.navigationDisabled}
          activeClassName={s.navigationActive}
        />
      </div>
    </>
  );
};
export default Main;
