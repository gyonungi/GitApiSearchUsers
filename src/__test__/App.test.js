import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import Main from "../pages/MainPage/Main";
import Card from "../components/Card/Card";

describe("TestApp", () => {
  test("input", () => {
    render(<Main />);
     const input = screen.getByPlaceholderText(/Поиск пользователя/i);
     expect(screen.queryByTestId("value-elem")).toContainHTML("");
     fireEvent.input(input,{
      target:{value:"123"}
     })
  });
});

test("click Event", () => {
  render(<Main />);
  const button = screen.getByText('Вниз');
  expect(button).toBeInTheDocument();
  const Ascbtn = screen.getByTestId("Asc-btn");
  const Descbtn = screen.getByTestId("Desc-btn");
  fireEvent.click(Ascbtn);
  fireEvent.click(Descbtn);
});


  test("renders listData Card", () => {
    render(<Card />);
    const list = screen.getAllByTestId("list-item");
    expect(list.length).toEqual(1);
  });

  test("use effect hooks" ,()=>{
    render(<Main/>)
    expect(screen.queryByText(/Не найден пользователь/i)).not.toBeInTheDocument()
  })