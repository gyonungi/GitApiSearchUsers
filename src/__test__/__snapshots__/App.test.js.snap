// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestApp input 1`] = `
<input
  class="searchText"
  name="search"
  placeholder="Поиск пользователя"
  type="text"
  value=""
/>
`;

exports[`TestApp renders learn react link 1`] = `
<input
  class="searchText"
  name="search"
  placeholder="Поиск пользователя"
  type="text"
  value=""
/>
`;
